feat: Add  MBRD and ROUGE score technique for majority voting

hickeyma opened the pull request:
[Minimum Bayesian Risk Decoding (MBRD)](https://arxiv.org/abs/2310.01387) is a technique that when used with [ROUGE ](https://en.wikipedia.org/wiki/ROUGE_(metric)) score similarity shows good accuracy when performing majority voting of model multi completion outputs per prompt.

This PR includes the following:

- Adds MBRD and ROUGE score algorithm
- Add new Python script to demonstrate how to use majority voting
- Add `MBRDMajorityVotingProcessor` composite processor

Closes: #72

frreiss reviewed (changes requested):
Thanks for doing this, but this implementation needs work.

The existing code for simple majority voting and the new code for minimum Bayesian risk decoding represent two of many methods for majority voting. Both of these methods have sub-variants depending on what types of normalizers, similarity scores, etc. the users plugs into them. There are other forms of majority voting that we do not currently have code for. For example, one might want to cluster the model outputs, then return the best representative of every cluster.

Majority voting techniques are supposed to be model-agnostic. They work with many different general-purpose LLMs, as well as with LLMs and LoRA adapters fine-tuned for specific applications and even with composite workflows that call multiple models. We shouldn't tie our implementations to a specific model or to inference operations that directly call a model as opposed to running a workflow.

The ramifications of the above to this code are as follows:
* We should not remove the existing code for simple majority voting. There are cases where that method is a better fit than MBRD. For example, one might want to return a set of top-k answers with diversity among the answers chosen.
* The MBRD code should be in its own composite I/O processor, so that it can be used with multiple different models. This composite I/O processor should allow the user to specify important parameters such as the choice of similarity metric.
* There should not be a Boolean "majority_voting" flag in the base Granite 3.2 model's I/O processor. There are many types of majority voting. Users should enable majority voting by configuring the appropriate composite I/O processor and attaching that I/O processor to the I/O processor of an LLM, LoRA adapter, or composite inference workflow.

hickeyma commented:
Thanks @frreiss for the feedback in https://github.com/ibm-granite/granite-io/pull/115#pullrequestreview-2737971493 and the discussion we had on best direction with supporting majority voting. I have taken your feedback on board and updated accordingly.

It is ready for review again.

markstur commented:
> > No idea whats up with DCO. Tried rebasing and force pushing but not fixing it.
> 
> one of the reverts is not signed off. Override should be okay here.

I set DCO to pass, but I'm not seeing the tests run.  Maybe it'd be best to do some squashing(?)

hickeyma commented:
>I set DCO to pass, but I'm not seeing the tests run. Maybe it'd be best to do some squashing(?)

All tests passed previously.

markstur commented on src/granite_io/io/granite_3_2/output_processors/granite_3_2_output_processor.py:
```diff
@@ -145,7 +146,7 @@ def transform(
                     next_message=AssistantMessage(
                         citations=parsed_output["citations"],
-                        content=parsed_output["response"],
+                        content=(parsed_output["response"]),
```
extraneous parens?

markstur commented on tests/io/test_voting.py:
```diff
@@ -65,3 +69,44 @@ def test_numeric_voting(backend_x: Backend):
     #         == first_number + second_number)
+
+
+@pytest.mark.vcr(record_mode="new_episodes")
```
We wouldn't want to commit with record_mode new_episodes.  I'd think this was just for creating/updating the cassette and should be turned off for CI just use the existing cassettes.

hickeyma replied:
Ok, changed for both tests in the file.

hickeyma replied:
Test failing when I remove the `record_mode new_episodes`.

markstur reviewed (changes requested):
2 little things inline that I think need fixing

In general, voting seems like internal experiments or examples more than a granite-io feature right now -- but I do see this as progress in the right direction.

hickeyma commented on examples/model_chat_with_majority_voting.py:
```diff
@@ -0,0 +1,46 @@
+print("------ WITH MAJORITY VOTING ------")
+# There will be only 1 output because majority voting is performed
+# on all completions results
+print(outputs.results[0].next_message.content)
```
The result should be 1 answer only. Yes, if you run with Ollama it will one have 1 completion regardless. The point is that you give the use of the script to configure an OpenAI backend that supports multiple completions that are used to give the majority result.

Updated now to:

```python
print("Output from base model augmented with MBRD majority voting:")
# This should be only one output, the majority voted answer
for result_num, r in enumerate(results.results):
    print(f"{result_num + 1}: {r.next_message.content}")
```

hickeyma commented on src/granite_io/io/granite_3_2/output_processors/granite_3_2_output_processor.py:
```diff
@@ -77,6 +80,16 @@ def transform(
         # This operation also performs additional validation.
         inputs = _Granite3Point2Inputs.model_validate(inputs.model_dump())
 
+        # Perform majority voting on sampled output
```
This is outdated and has been changed to be independent of the model IO processor.

markstur commented:
So sorry... updating with main was an accident.

frreiss commented on src/granite_io/io/granite_3_2/output_processors/granite_3_2_output_processor.py:
```diff
@@ -157,3 +158,24 @@ def transform(
         return ChatCompletionResults(results=results)
+
+
+def extract_normalized_responses(results: GenerateResults, tag: str) -> list[str]:
```
This function doesn't seem to be called anywhere?

hickeyma replied:
Good catch. Removed in #139

frreiss commented on src/granite_io/io/voting/__init__.py:
```diff
@@ -3,6 +3,7 @@
 
 # Local
+from granite_io.io.voting.mbrd_voting import MBRDMajorityVotingProcessor  # noqa: F401
 from granite_io.io.voting.voting import (  # noqa: F401
```
Maybe rename these two files to `mbrd.py` and `simple.py`

hickeyma replied:
Makes sense. Renamed in #139

frreiss commented on src/granite_io/io/voting/mbrd_voting.py:
```diff
@@ -0,0 +1,97 @@
+    """
+
+    similarity_scores: list[float] = []
+    for _, x in enumerate(answers):
```
Should be `for x in answers`, not sure why the linter didn't catch this.

hickeyma replied:
Not sure what you mean by this. The `enumnerate()` function will return counter and object (string in this case). If you don't account for the counter you end up with tuples like this: `(0, 'str1') (1, 'str2') ....`

frreiss commented on src/granite_io/io/voting/mbrd_voting.py:
```diff
@@ -0,0 +1,97 @@
+
+
+def _minimum_bayesian_risk_decoding(
+    answers: list[str], similarity: Callable
```
What's the signature of `similarity`? Is it `Callable[[str],float]`, returning a float between 0. and 1. inclusive, or something else?

hickeyma replied:
Yes, updated to add the signature in #139

frreiss commented on tests/io/cassettes/test_voting/test_mbrd_majority_voting[backend_openai].yaml:
```diff
@@ -0,0 +1,66 @@
+    method: POST
+    uri: http://localhost:11434/v1/completions
+  response:
+    content: '{"id":"cmpl-153","object":"text_completion","created":1743878084,"model":"granite3.2:2b","system_fingerprint":"fp_ollama","choices":[{"text":"Here
```
Only one completion here. Was this test run against a backend that doesn't support returning multiple answers?

markstur replied:
yes we are typically using Ollama (http://localhost:11434) and it returns 1.

I think we should all move to vllm, but would like us in sync on that.

hickeyma replied:
Ok, lets handle in a separate PR for all tests.

frreiss reviewed (approved):
Looks good. Some minor comments inline.

markstur reviewed (approved):
Fred added comments that you might want to address, but otherwise you have approvals to merge.

hickeyma commented:
Thanks @frreiss and @markstur for the reviews and feedback. 

@frreiss I have addressed your feedback from https://github.com/ibm-granite/granite-io/pull/115#pullrequestreview-2754844705 in #139.
